#211 Simplified project and build system.

niemyjski opened the pull request:
References #211 

NOTE: I've yet to run any of the tests but I made very minimal changes to the tests and code.

1. I updated to .NET Standard 2.0 from 1.6 as the latest version of RestSharp only supports .NET Standard 2.0. It would be nice if we could drop .NET 4.5.2 in the near future and just target .NET standard 2.0, this simplifies the cross platform story.
2. I Dropped any unused dependencies, duplicated test classes and the file upload project (as it seemed like didn't do a single thing that the tests weren't already doing).
3. For time sake, I just removed all System.Configuration references and hard coded the public play info. It would be nice to add configuration for this later.
4. The Chinese readme may need some work, I tried the best I could to update it. (I only speak English)
5. All projects are now strong name signed, this has no impact and allows more people to use the library, it also fixed internalsvisibleto issues with unit tests that all now run on .net core :)

niemyjski commented:
@poornas I just noticed that the build is failing because newly discovered unit tests in the `UnitTest1.cs` are failing. It looks like not all the test classes were included before so I've marked them as skip.. Here is the output showing which test classes were discovered before: `DictionaryExtensionMethods.cs NotificationTest.cs PolicyConditionTests.cs Properties\AssemblyInfo.cs EndpointTest.cs TestHelper.cs PolicyTests.cs AuthenticatorTest.cs RegionTest.cs UtilsTest.cs UrlTests.cs `

poornas commented:
Thanks @niemyjski, the chinese readme can be updated later - the original contributors to the chinese docs might assist with the changes. The .net4.5.2 package has had ~3800 downloads in the last few months, so for now it seems good to retain .net4.5.2 support.

niemyjski commented:
@poornas all tests and builds are now passing!. However travis tests look like they are having some failures outputted to the console. I didn't change anything in any of the tests / classes just very minor things so any ideas what those are about? I also enabled automatic version patching in AppVeyor that you can play around with just set patch to true.

niemyjski commented:
@poornas I changed the user agent to be the current version was 0.1.2 (this needs to be automated..

niemyjski commented:
Can you please review this pr :)

poornas commented:
@niemyjski, pulled in your changes and tested on Linux - the signature mismatch error needs to be fixed. Our automated [tests](https://github.com/minio/mint) pull in functional tests from the sdk and test this in a docker container using a linux image.Will test this tonight on Visual Studio

poornas commented on FileUploader/FileUpload.cs:
```diff
@@ -1,83 +0,0 @@
-﻿/*
- * Minio .NET Library for Amazon S3 Compatible Cloud Storage, (C) 2017 Minio, Inc.
```
The FileUploader project is on the SDK to give a quick running example to new/inexperienced users to demo basic functionality - imo, we should keep this example.

niemyjski replied:
@poornas I can add it back, I didn't think it was adding any value and was just noise. We already  have 3 test projects and this should be combined down to one. This should go in a test or snippet someplace else. There are already sample code in the test project that contains every operation.

niemyjski replied:
Thoughts on this?

niemyjski commented:
@poornas the minio.snk doesn't need to be removed. It's for strong name signing of the assembly which allows the assemblies to be used with or without signing. It's literally a joke and should have been removed from .net core. Almost every major project includes there keys and it doesn't hurt anything, just a progressive enhancement. And when I say this, I can open a hex editor / some tool and give any assembly a different strong name key or remove it so it's really not harming anyone.

niemyjski commented:
@poornas do you have any ideas on fixing the tests signatures? Mind submitting a pr?

poornas commented:
@niemyjski , the AWS s3 reference for constructing authorization header is [here](https://docs.aws.amazon.com/AmazonS3/latest/API/sig-v4-header-based-auth.html).  The authorization header is computed in UTF-8 whereas C# uses UTF-16 -  when we extract request parameters to construct authorization header, some extraneous byte(s) are not getting encoded properly or are getting double encoded. As the only variable here is the Restsharp switch from Restsharp.Netcore, it seems to be a encoding difference while extracting parameters from the request. 

 The AWS C# signature calculation [example](https://docs.aws.amazon.com/AmazonS3/latest/API/sig-v4-examples-using-sdks.html#sig-v4-examples-using-sdk-dotnet) can help debug the issue. When you tested, did the SimpleTest.csproj list buckets at all?

niemyjski commented:
I just added the FileUploader sample back and also took a quick look. I don't have time to dive into this but I can tell you it's simply using the `request.AddHeader("x-amz-content-sha256", "e3b0c44298fc1c149afbf4c8996fb92427ae41e4649b934ca495991b7852b855");` call in V4 Authenticator. Can you please create a new branch in the main repo and merge this in and look into this. This is not something I changed other than running it under .net core (perhaps) which means this is preexisting.

poornas commented:
@niemyjski, will check the signature mismatch issue this weekend and send a PR to your branch. The issue seems more related to swapping Restsharp.Netcore out with Restsharp rather than a .netcore 2.0 issue, but will have to dig in further to isolate the problem.

poornas commented:
@niemyjski , Restsharp is stripping off the port in the Host header causing a signature mismatch. The minio sdk uses the full hostname with port to calculate the authorization signature. Since this is stripped off , the signature calculated by server doesn't match the signature string sent by the sdk.

niemyjski commented:
@poornas do you want to open an issue for restsharp on this? or can we manually add it? We were on 105 and we went to 106 but I thought the only changes were netstandard support. You also have to take into account we were using an unofficial port for .net core and we have no idea what changed in it as the source isn't available.

poornas commented:
marking this as blocked until Restsharp issue is resolved.

niemyjski commented:
I think I resolved the issues with this pr. I need to finish with the merge conflicts and push. I'll try to do that in the morning.

twgraham commented:
How'd you go @niemyjski ? I'm super keen for this PR 😃

niemyjski commented:
I need to get to this today. It's just been really crazy.

niemyjski commented:
I pulled in latest changes, lets see if the tests pass via the ci :)

niemyjski commented:
@poornas can you please review the changes and merge the pr :)

poornas commented:
@niemyjski , can you squash your commits to a single commit for easier review?

niemyjski commented:
I just tried real quick and I was only able to squash 4 commits (the ones after I merged in origin because rebases was a pain)

poornas commented:
get a bunch of errors similar to the ones below while compiling on VS or Linux - please fix these. 
```
  Minio.Client.Examples.Core/obj/Debug/netcoreapp1.1/Minio.Client.Examples.Core.AssemblyInfo.cs(14,12): error CS0579: Duplicate 'System.Reflection.AssemblyFileVersionAttribute' attribute [/home/kris/code/minio-dotnet/Minio.Examples/Minio.Examples.csproj]
  Minio.Client.Examples.Core/obj/Debug/netcoreapp1.1/Minio.Client.Examples.Core.AssemblyInfo.cs(15,12): error CS0579: Duplicate 'System.Reflection.AssemblyInformationalVersionAttribute' attribute [/home/kris/code/minio-dotnet/Minio.Examples/Minio.Examples.csproj]
  Minio.Client.Examples.Core/obj/Debug/netcoreapp1.1/Minio.Client.Examples.Core.AssemblyInfo.cs(16,12): error CS0579: Duplicate 'System.Reflection.AssemblyTitleAttribute' attribute [/home/kris/code/minio-dotnet/Minio.Examples/Minio.Examples.csproj]
  Minio.Client.Examples.Core/obj/Debug/netcoreapp1.1/Minio.Client.Examples.Core.AssemblyInfo.cs(17,12): error CS0579: Duplicate 'System.Reflection.AssemblyVersionAttribute' attribute [/home/kris/code/minio-dotnet/Minio.Examples/Minio.Examples.csproj]

```

poornas commented:
@niemyjski , can you add back the configuration for Mint for the Functional tests project - it was referencing the version of Minio.Netcore published on nuget for our automated testing. The Release.Core and Release.Net452 configurations were in place so that the dotnet core and .net framework code would compile on linux. Did you test whether .net452 projects in the solution will compile with ```dotnet build \p:Configuration=Release``` on linux?

niemyjski commented:
@poornas why do you want the snk removed? It's 100% pointless to remove it. There are hundreds of libs including json.net that ship the snk. Also no I didn't those if they weren't on travis.. the tests run on .net core 2.1 which works on cross platform on every platform. So I don't know why we would need a custom target for this... Also you are running into that build error because you don't have the latest sdk or runtime (v2.1) on your machine. You are trying to build with the 1.1 runtime I believe which will error.

niemyjski commented:
Let me know what else I need to do. The tests are passing on travis so I'm not sure what else is required.

twgraham commented:
Lets 👏 get 👏 this 👏 in 👏  😄 . @poornas whats the update with regards to @niemyjski's comment. @niemyjski if the snk is the only thing keeping this from being merged, lets just get rid of it. This is some good stuff!

poornas commented on Minio.Functional.Tests/Minio.Functional.Tests.csproj:
```diff
@@ -1,43 +1,21 @@
-            <ProjectReference Include="..\Minio.Core\Minio.Core.csproj" />
-          </ItemGroup>
-        </When>
-        <When Condition=" '$(Configuration)'=='Mint' ">   
```
@niemyjski , can you leave the Mint Configuration here as is - our automated tests need to reference the published version of Minio.NetCore

niemyjski replied:
I'll get this in but I'm unable to rebase, it's complaining when I try, you are able to squash and merge when merging the pr (it's a drop down option)

harshavardhana commented on Minio/V4Authenticator.cs:
```diff
@@ -71,9 +71,10 @@ internal class V4Authenticator : IAuthenticator
         /// <param name="accessKey">Access key id</param>
         /// <param name="secretKey">Secret access key</param>
+        /// <param name="region"></param>
         public V4Authenticator(bool secure,string accessKey, string secretKey, string region ="us-east-1")
```
This API has been changed region is not available anymore, we can remove the reference now?

harshavardhana commented:
@niemyjski - Thanks for the commendable effort here, I only have one comment. Other than that we are good to proceed with merging this to master and making a new release. 

Additionally, can you change the version and let's move to 2.0.0 with this change? what do you think @poornas ?

poornas commented:
@harshavardhana, agree that we should bump up the version, since the SDK is moving up to .NET standard 2.0

niemyjski commented:
I'll also bump the version and get the mint config in.

poornas commented on Minio.core.nuspec:
```diff
@@ -1,31 +0,0 @@
-  <files>
-      <file src="Minio.Core\bin\Release.Core\**" target="lib" />
-  </files>
-</package>
```
@niemyjski, this file is needed for creating nuget package - can you change the target framework, version (2.0.0) and update dependencies.

poornas commented on Minio.nuspec:
```diff
@@ -1,35 +0,0 @@
-  <files>
-      <file src="Minio.Net452\bin\Release.Net452\**" target="lib" />
-  </files>
-</package>
```
@niemyjski, MInio.nuspec is needed for creating nuget package by our automated release system. Can you add this file back with updated version, dependencies and target framework

niemyjski replied:
No, these are not needed, the new build system will automatically create packages for you. We shouldn't reinvent the wheel here. The build is creating nuget packages that match the previous system with tons of deleted code == easier to maintain.

poornas commented:
@harshavardhana , we need the Mint config change and the Minio.nuspec and Minio.core.nuspec to be added back before merging.Looks great @niemyjski , thanks for all the effort.

niemyjski commented:
I've made all the requested changes including the snk file, please squash on merge (option). Also unspec files are not needed, packages are already being created exactly how they were via the appveyor build.

poornas commented:
@niemyjski, our automated release builds the package on linux, that's why the nuspec is needed.

niemyjski commented on appveyor.yml:
```diff
@@ -1,30 +1,39 @@
+  - ps: dotnet build -c Release
+
+after_build:
+  - ps: dotnet pack .\Minio\Minio.csproj --no-build -c Release -o .\artifacts
```
@poornas all you gotta do is run this command and it will automatically package the nuget packages on any platform (osx, windows, linux)

poornas replied:
ok.

niemyjski commented:
Please see my review comment on how to properly build nuget packages on any platform using the built in dotnet cli tooling... one should not be using the old way with nuspec files. There is only one rare edge case where you'd want to use it and that's dealing with legacy project types (WPF) on no longer supported framework (< .NET 4.5)...

harshavardhana commented:
> @niemyjski, our automated release builds the package on linux, that's why the nuspec is needed.

We need to make these changes to the automated build without nuspec @poornas

poornas commented:
@harshavardhana , will do that once this is merged

niemyjski commented:
Thanks!
